src/main/sensors/barometer.c

```diff
@@ -322,28 +324,43 @@ bool isBaroReady(void) {
 
 uint32_t baroUpdate(void)
 {
-    static barometerState_e state = BAROMETER_NEEDS_SAMPLES;
+    static barometerState_e state = BAROMETER_NEEDS_PRES_START;
 
     switch (state) {
         default:
-        case BAROMETER_NEEDS_SAMPLES:
+        case BAROMETER_NEEDS_TEMP_START:
+            baro.dev.start_ut(&baro.dev);
+            state = BAROMETER_NEEDS_TEMP_SAMPLE;
+            return baro.dev.ut_delay;
+            break;
+
+        case BAROMETER_NEEDS_TEMP_SAMPLE:
             baro.dev.get_ut(&baro.dev);
+            state = BAROMETER_NEEDS_PRES_START;
+        break;
+
+        case BAROMETER_NEEDS_PRES_START:
             baro.dev.start_up(&baro.dev);
-            state = BAROMETER_NEEDS_CALCULATION;
+            state = BAROMETER_NEEDS_PRES_SAMPLE;
             return baro.dev.up_delay;
         break;
 
-        case BAROMETER_NEEDS_CALCULATION:
+        case BAROMETER_NEEDS_PRES_SAMPLE:
             baro.dev.get_up(&baro.dev);
-            baro.dev.start_ut(&baro.dev);
+
             baro.dev.calculate(&baroPressure, &baroTemperature);
             baro.baroPressure = baroPressure;
             baro.baroTemperature = baroTemperature;
             baroPressureSum = recalculateBarometerTotal(barometerConfig()->baro_sample_count, baroPressureSum, baroPressure);
-            state = BAROMETER_NEEDS_SAMPLES;
-            return baro.dev.ut_delay;
+            if (baro.dev.combined_read) {
+            	state = BAROMETER_NEEDS_PRES_START;
+            } else {
+            	state = BAROMETER_NEEDS_TEMP_START;
+            }
         break;
     }
+
+    return SCHEDULER_DELAY_LIMIT;
```

Comment:
The intention is to split the baro task into separately schedulable parts, but to execute them as quickly as possible. The behaviour is exactly as intended. SCHEDULER_DELAY_LIMIT is the correct value to return.

Reply:
There is unfortunately a fundamental issue here. I've modified the scheduler to calculate the amount of remaining time before the next realtime task must run. It then only runs tasks that have a maxExecutionTime less that this. This fixes your concern above, so I can return SCHEDULER_DELAY_LIMIT and know that the scheduler will only schedule the next task execution when it has time to do so. 

It also records the peak amount of available time it ever sees. On my FC this is 81us. I've timed the baro access phases.

| State  | Duration |
| -----  | -------- |
| BAROMETER\_NEEDS\_TEMP_SAMPLE | 82 |
| BAROMETER\_NEEDS\_PRES_START | 43 |
| BAROMETER\_NEEDS\_PRES_SAMPLE | 86 |
| BAROMETER\_NEEDS\_TEMP_START  | 44 |
| BAROMETER\_NEEDS\_CALC | 4 |

So, unfortunately the I2C bus access is taking too long, the I2C accesses taking longer than the loop time will allow. The only way to fix this will be to take advantage of the interrupt driven nature of the I2C bus driver so the task can return immediately and leave the I2C access to run in the background. This will mean modifying the bus driver interface so that i2cRead (or a new version thereof) can start the transfer, and return immediately and that another function can be called to test for completion.

I'll code this up and see how well it works. This will have the advantage of freeing up some processor time.

Reply:
Please, let's focus on resolving the task at hand and splitting the baro read phases before going off on additional tangents (that should not be included in this PR). No scheduler changes, no changes to I2C - save this for other PR's as they're developed. We don't want to introduce variability into the baro read by trying to predict against other tasks in the scheduler.

Reply:
In order for this baro PR to be meaningful it needs those changes, so I propose raising PRs for the I2C bus and scheduler changes and then returning to this one.